test/test_tutorial.cpp

```diff
@@ -170,4 +170,280 @@ TEST_F(Tutorial, Memcpy) {
+    std::vector<at::Tensor> outputs = fe.runFusion({t1});
+    testValidate(
+        &fusion, outputs, aten_inputs, {t1.sum({1})}, __LINE__, __FILE__);
+  }
+
+  // We can of course mix BIDx and TIDx.
+  tv1->axis(0)->parallelize(ParallelType::BIDy);
+  tv1->axis(1)->parallelize(ParallelType::TIDx);
+
+  if (verbose_) {
+    fusion.printMath();
+    fusion.printKernel();
+  }
+
+  {
+    FusionExecutor fe;
+    fe.compileFusion(&fusion);
+    // The original input should not fail in this case. The kernel
+    // will be launched with 10 thread blocks, each of which has 1024
+    // threads. Try running this test with NVFUSER_DUMP=launch_param
+    // to see the launch configuration of each kernel lauch
+    std::vector<at::Tensor> outputs = fe.runFusion(aten_inputs);
+    testValidate(&fusion, outputs, aten_inputs, {ref}, __LINE__, __FILE__);
+  }
+}
+
+TEST_F(Tutorial, ReductionRFactor) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  // Just a very simple reduction of 1D tensor
+  TensorView* tv0 = makeSymbolicTensor(1);
+  fusion.addInput(tv0);
+  auto tv1 = sum(tv0, {0});
+  fusion.addOutput(tv1);
+
+  if (verbose_) {
+    fusion.printMath();
+  }
+
+  // Multiple scheduling examples will be created so each time a copy
+  // of the fusion is used
+  {
+    Fusion fusion_copy = fusion;
+    FusionGuard fg_copy(&fusion_copy);
+
+    // Note that tv1 is not a valid pointer for fusion_copy
+    auto tv1_copy = fusion_copy.outputs().at(0)->as<TensorView>();
+
+    // A common pattern of reductions in CUDA involves multiple steps of
+    // reductions, where the first step is a per-thread local reduction,
+    // followed by a block reduction of the per-thread partial results,
+    // and also potentially followed by a grid reduction of the
+    // per-block partial results. Here's an example with a two-step
+    // reduciton:
+    //
+    // // Step 1: Per-thread reduction
+    // float partial_result = 0;
+    // for (int i = threadIdx.x; i += blockDim.x; i < N) {
+    //   partial_result += input[i];
+    // }
+    //
+    // // Step 2: Accumulation within each thread block
+    // __shared__ float shared_buf[blockDim.x];
+    // shared_buf[threadIdx.x] = partial_result;
+    // __syncthreads();
+    // float final_result = 0;
+    // // Accumulation of the partila result in a naive sequntial way.
+    // if (threadIdx.x == 0) {
+    //   for (int i = 0; i < blockDim.x; ++i) {
+    //     final_result += shared_buf[i];
+    //   }
+    // }
+    //
+    // To reproduce the multi-step reduction pattern in nvFuser, a
+    // fusion transformaiton called reduction rfactor is used. The basic
+    // idea is to split a reduction domain such that each of the output
+    // domains of the split is separately reduced. For example, tv1 can
+    // be transformed from a 2D tensor to a 3D tensor as follows:
+    //
+    // tv0: [i0]
+    // tv1: [r1]
+    tv1_copy->split(0, 1024);
+    // tv1: [r1/1024, r1024]
```

Comment:
OOC, is `r1024` named intentionally? I thought it would be `r2` (using a running unique ID) with extent of `1024`. I didn't expect the ID to be bumped up that high immediately.

Reply:
Yes, the actual name should be something like `r2`. I just used `1024` to indicate its extent.